far/vmenu.cpp

```diff
@@ -140,15 +140,45 @@ void VMenu::ResetCursor()
 	GetCursorType(PrevCursorVisible,PrevCursorSize);
 }
 
+template <typename coll, typename pred>
+int FindNearestSelectableItem(const coll& Coll, const int Pos, pred&& Pred, const bool GoBackward, const bool DoWrap)
+{
+	assert(0 <= Pos && Pos < static_cast<int>(Coll.size()));
+
+	const auto FindPos{
+		[&](const auto First, const auto Second)
+		{
+			const auto FindPos{
+				[&](const auto Part)
+				{
+					const auto Found{ std::ranges::find_if(Part, Pred) };
+					if (Found == std::ranges::end(Part)) return -1;
+					return static_cast<int>(std::ranges::distance(&*std::ranges::begin(Coll), &*Found));
+				} };
+
+			if (const auto Found{ FindPos(Coll | First) }; Found != -1) return Found;
+			if (const auto Found{ FindPos(Coll | Second) }; Found != -1) return Found;
+			return -1;
+		} };
+
+	return GoBackward
+		? (DoWrap
+			? FindPos(std::views::take(Pos + 1) | std::views::reverse, std::views::drop(Pos + 1) | std::views::reverse)
```

Comment:
`using namespace std::views;` is, of course, the ticket. I thought about it but at the last moment forgot to drop it in. There are places where people routinely use fully qualified `std` identifiers and do not think they're mouthful.

As an aside, I hesitated to go full ranges, partly because some features could be unsupported by all compilers. To be honest, I was lazy to try using every little thing formally available in C++20 and only fall back if some compilers complain.

P.S. Maybe it's time to go C++23? 🙂

Reply:
According to [cppreference.com](https://en.cppreference.com/w/cpp/compiler_support#C.2B.2B20_library_features), ranges are supported in:

GCC|Clang|MSVC|
-- | -- | --
10 | 13 (partial) 15 | 19.29 (16.10)

Our current requirements are:
https://github.com/FarGroup/FarManager/blob/5c9bce3ac7bb97fbd81c40d1c8e4c9ce35ac7d84/far/headers.hpp#L40-L46

So it should be ok.

Reply:
Looks like everyone's happy, so I am.